speaker/bgp_controller.go

```diff
@@ -272,20 +391,264 @@ type session interface {
+			case pad.FromLabels.Addr:
+				peerAddr = net.ParseIP(v)
+				if peerAddr == nil {
+					return nil, fmt.Errorf("invalid peer IP %q", v)
+				}
+			case pad.FromLabels.Port:
+				port, err := strconv.ParseUint(v, 10, 16)
+				if err != nil {
+					return nil, fmt.Errorf("parsing peer port: %v", err)
+				}
+				peerPort = uint16(port)
+			case pad.FromLabels.HoldTime:
+				holdTimeRaw = v
+			case pad.FromLabels.RouterID:
+				routerID = net.ParseIP(v)
+				if routerID == nil {
+					return nil, fmt.Errorf("invalid router ID %q", v)
+				}
+			}
+		}
+	}
+
+	if pad.FromAnnotations != nil {
+		for k, v := range anns {
+			switch k {
+			case pad.FromAnnotations.MyASN:
+				asn, err := strconv.ParseUint(v, 10, 32)
+				if err != nil {
+					return nil, fmt.Errorf("parsing local ASN: %v", err)
+				}
+				myASN = uint32(asn)
+			case pad.FromAnnotations.ASN:
+				asn, err := strconv.ParseUint(v, 10, 32)
+				if err != nil {
+					return nil, fmt.Errorf("parsing peer ASN: %v", err)
+				}
+				peerASN = uint32(asn)
+			case pad.FromAnnotations.Addr:
+				peerAddr = net.ParseIP(v)
+				if peerAddr == nil {
+					return nil, fmt.Errorf("invalid peer IP %q", v)
+				}
+			case pad.FromAnnotations.Port:
+				port, err := strconv.ParseUint(v, 10, 16)
+				if err != nil {
+					return nil, fmt.Errorf("parsing peer port: %v", err)
+				}
+				peerPort = uint16(port)
+			case pad.FromAnnotations.HoldTime:
+				holdTimeRaw = v
+			case pad.FromAnnotations.RouterID:
+				routerID = net.ParseIP(v)
+				if routerID == nil {
+					return nil, fmt.Errorf("invalid router ID %q", v)
+				}
+			}
+		}
+	}
+
+	// Verify required peer config. We shouldn't get errors here because we
+	// validate the configuration. This check is here just for safety.
+	if myASN == 0 {
+		return nil, errors.New("missing local ASN")
+	}
+	if peerASN == 0 {
+		return nil, errors.New("missing peer ASN")
+	}
+	if peerAddr == nil {
+		return nil, errors.New("missing peer address")
+	}
+
+	// Set default BGP port if unspecified by user.
+	if peerPort == 0 {
+		peerPort = 179
+	}
+
+	if holdTime == 0 {
+		// Hold time not specified in autodiscovery defaults - try to parse the
+		// hold time from labels/annotations.
+		ht, err := parseHoldTime(holdTimeRaw)
+		if err != nil {
+			return nil, fmt.Errorf("parsing hold time: %v", err)
+		}
+		holdTime = ht
+	}
+
+	// The peer is configured on a specific node object, so we want to create a
+	// BGP session only on that node.
+	h := ls[v1.LabelHostname]
+	if h == "" {
+		return nil, fmt.Errorf("label %s not found on node", v1.LabelHostname)
+	}
+	ns, err := labels.Parse(fmt.Sprintf("%s=%s", v1.LabelHostname, h))
+	if err != nil {
+		return nil, fmt.Errorf("parsing node selector: %v", err)
+	}
+
+	p := &peer{
+		Cfg: &config.Peer{
+			MyASN:         myASN,
+			ASN:           peerASN,
+			Addr:          peerAddr,
+			Port:          peerPort,
+			HoldTime:      holdTime,
+			RouterID:      routerID,
+			NodeSelectors: []labels.Selector{ns},
```

Comment:
The `syncBGPSessions()` function is used for reconciling both regular peers and node peers. This function relies on the `NodeSelectors` field to determine whether a peer should be active for a node. Rather than creating a separate function for node peers, I've opted for adopting the existing logic and creating node peers with a node selector which means "this node". I'm willing to consider alternatives if you have one you think is preferred.